ouroboros-network/src/Ouroboros/Network/DeltaQ.hs

```diff
@@ -242,13 +242,48 @@ gsvTrailingEdgeArrive (GSV g s v) bytes =
 -- | The 'GSV' for both directions with a peer, outbound and inbound.
 --
 data PeerGSV = PeerGSV {
+                 sampleTime  :: !Time,
                  outboundGSV :: !GSV,
                  inboundGSV  :: !GSV
                }
                deriving Show
 
+-- | The current tracking model is based on an EWMA
+--  (https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average).
+--  Typically implementations of EWMA assume a regular update, but EWMA is based
+--  on Exponential Smoothing
+--  (https://en.wikipedia.org/wiki/Exponential_smoothing). Such smoothing has a
+--  time constant, which captures the time for a unit impulse to decay to 1 -
+--  1/e (~ 63.2%), the &#x1D6FC (smoothing factor) is a function of relative
+--  frequency of the sample interval and this time constant.
+--
+-- The approach being taken here is one that does not assume a fixed sample
+-- interval (and hence a fixed &#x1D6FC), instead we calculate, given the
+-- interval from when the last sample was taken, the &#x1D6FC needed to ensure
+-- that the old value has sufficiently decayed.
+--
+-- The exact calcuation involves exponentiation, however where the number of
+-- samples within the time constant is sufficiently large a simple ratio of the
+-- sample's interval over the time constant will suffice. The relative error of
+-- this numerical approximation is, for our use case, small. Eg 1/50 (20s
+-- between samples with a 1000s time constant) has a relative error of 1%. The
+-- expected typical range of this relative error is between 5% (ratio of 1/10),
+-- to 0.5% (1/100).
+--
+-- Given the inherent measurement noise in this measurement, the use of the
+-- approximation is well justified. We choose (reaonably aribtarily) 1000s as
+-- the time constant, it is unclear if this should be a configuration variable
+-- or not.
```

Comment:
I would add a note that this `Semigroup` is non-commutative.

The new value must come first, which is not the order that is used in the `keepAliveClient`.

Reply:
Will add it to the comment.

Reply:
I think we also need to fix it [here](https://github.com/input-output-hk/ouroboros-network/pull/2251/files#diff-e8529b129ff6f1539758a2ef6947310bR83)

Reply:
Dang, you're right!

Reply:
:1st_place_medal:

Reply:
I wonder how this affects your tests ...